docs/source/miscellaneous/doc/adr/0015-batch-asynchronous-post-transaction-unlock-calls.md

```diff
@@ -0,0 +1,145 @@
+
+In our implementation of solution 2, we use a single-threaded executor. This means that on average the additional
+latency we incur is about 0.5 RPCs on the lock service (assuming that that makes up a majority of time spent in
+unlocking tokens - it is the only network call involved).
+
+### tryUnlock() API
+
+`TimelockService` now exposes a `tryUnlock()` API, which functions much like a regular `unlock()` except that the user
+does not need to wait for the operation to complete. This API is only exposed in Java (not over HTTP).
+
+This is implemented as a new default method on the `TimelockService` that delegates to `unlock()`; usefully, remote
+Feign proxies calling `tryUnlock()` will make an RPC for standard `unlock()`. This also gives us backwards
+compatiblity; a new AtlasDB/TimeLock client can talk to an old TimeLock server that has no knowledge of this endpoint.
+
+### Concurrency Model
+
+It is essential that adding an element to the set of outstanding tokens is efficient; yet, we also need to ensure that 
+no token is left behind (at least indefinitely). We thus guard the concurrent set by a (Java) lock that permits both 
+exclusive and shared modes of access.
+
+Transactions that enqueue lock tokens to be unlocked perform the following steps:
+
+1. Acquire the set lock in shared mode.
+2. Read a reference to the set of tokens to be unlocked.
+3. Add lock tokens to the set of tokens to be unlocked.
+4. Release the set lock.
+5. If no task is scheduled, then schedule a task by setting a 'task scheduled' boolean flag. 
+   This uses compare-and-set, so only one task will be scheduled while no task is running.
+
+For this to be safe, the set used must be a concurrent set. 
+
+The task that unlocks tokens in the set performs the following steps:
+
+1. Un-set the task scheduled flag.
+2. Acquire the set lock in exclusive mode.
+3. Read a reference to the set of tokens to be unlocked.
+4. Write the set reference to point to a new set. 
+5. Release the set lock.
+6. Unlock all tokens in the set read in step 3.
+
+This model is trivially _safe_, in that no token that wasn't enqueued can ever be unlocked, since all tokens that can
+ever become unlocked must have been added in step 3 of enqueueing, and unlocking a lock token is idempotent modulo
+a UUID clash.
+
+More interestingly, we can guarantee _liveness_ - every token that was enqueued will be unlocked in the absence of
+thread death. If an enqueue has a successful compare-and-set in step 5, then the token must be in the set
+(and is visible, because we synchronize on the set lock). If an enqueue does _not_ have a successful compare-and-set,
+then some thread must already be scheduled to perform the unlock, and once it does the token must be in the relevant
+set (and again must be visible, because we synchronize on the set lock).
+
+To avoid issues with starving unlocks, we use a fair lock scheme. This may have lower throughput than an unfair lock,
```

Comment:
This could use elaboration, e.g. that enqueues and unlocks are given the set lock in the order they requested it, rather than enqueues having priority.